.gitlab-ci.yml

```diff
@@ -1120,6 +1139,166 @@ workflow:
   - when: manual
     allow_failure: true
 
+.on_asc_or_e2e_changes_or_manual:
+  - <<: *if_disable_e2e
+    when: never
+  - <<: *if_auto_e2e
+    when: on_success
+  - <<: *if_main_branch
+    when: on_success
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
Noticed we used both `always` and `on_success` for main branch, are you using `on_success` for when `needs` is not defined ?

Reply:
I just realized we had both. I'd prefer to put `on_success` everywhere. Because with `always` you can end with the job being executed while its dependencies failed